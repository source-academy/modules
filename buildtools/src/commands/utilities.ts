/* [Exports] */
export function degreesToRadians(degrees: number): number {
  return (degrees / 360) * (2 * Math.PI);
}

export function hexToColor(hex: string): [number, number, number] {
  const regex = /^#?([\da-f]{2})([\da-f]{2})([\da-f]{2})$/igu;
  const groups = regex.exec(hex);

  if (groups == undefined) return [0, 0, 0];
  return [
    parseInt(groups[1], 16) / 0xff,
    parseInt(groups[2], 16) / 0xff,
    parseInt(groups[3], 16) / 0xff
  ];
}
