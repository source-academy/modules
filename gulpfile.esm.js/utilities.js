/**
 * Utilities for scripts
 */
import { dirname } from 'path';
import { fileURLToPath } from 'url';

// eslint-disable-next-line import/prefer-default-export
export const cjsDirname = () => {
  return dirname(fileURLToPath(import.meta.url));
};
