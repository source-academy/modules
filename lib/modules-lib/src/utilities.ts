import type { DebuggerContext } from './types';

/* [Exports] */
export function degreesToRadians(degrees: number): number {
  return (degrees / 360) * (2 * Math.PI);
}

export function hexToColor(hex: string): [number, number, number] {
  const regex = /^#?([\da-f]{2})([\da-f]{2})([\da-f]{2})$/igu;
  const groups = regex.exec(hex);

  if (groups == undefined) return [0, 0, 0];
  return [
    parseInt(groups[1], 16) / 0xff,
    parseInt(groups[2], 16) / 0xff,
    parseInt(groups[3], 16) / 0xff
  ];
}

export function mockDebuggerContext<T>(state: T, module: string) {
  return {
    context: {
      [module]: {
        state
      }
    }
  } as unknown as DebuggerContext;
}
