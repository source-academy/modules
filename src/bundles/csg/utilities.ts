/* [Imports] */
import geom3, {
  transform as _transform
} from '@jscad/modeling/src/geometries/geom3';
import mat4, { type Mat4 } from '@jscad/modeling/src/maths/mat4';
import {
  center as _center,
  rotate as _rotate,
  scale as _scale,
  translate as _translate
} from '@jscad/modeling/src/operations/transforms';
import type { ReplResult } from '../../typings/type_helpers';
import { Core } from './core';
import type { AlphaColor, Color, Solid } from './jscad/types';

/* [Exports] */
export interface Operable {
  applyTransforms: (newTransforms: Mat4) => Operable;
  store: (newTransforms?: Mat4) => void;

  translate: (offsets: [number, number, number]) => Operable;
  rotate: (angles: [number, number, number]) => Operable;
  scale: (factors: [number, number, number]) => Operable;
}

export class Group implements Operable, ReplResult {
  children: Operable[];

  constructor(_children: Operable[], public transforms: Mat4 = mat4.create()) {
    // Duplicate the array to avoid modifying the original, maintaining
    // stateless Operables for the user
    this.children = [..._children];
  }

  applyTransforms(newTransforms: Mat4): Operable {
    const appliedTransforms: Mat4 = mat4.multiply(
      mat4.create(),
      newTransforms,
      this.transforms
    );

    // Return a new object for statelessness
    return new Group(this.children, appliedTransforms);
  }

  store(newTransforms: Mat4 = mat4.create()): void {
    const appliedGroup: Group = this.applyTransforms(newTransforms) as Group;

    this.children.forEach((child: Operable) => {
      child.store(appliedGroup.transforms);
    });
  }

  translate(offsets: [number, number, number]): Group {
    return new Group(
      this.children,
      mat4.multiply(
        mat4.create(),
        mat4.fromTranslation(mat4.create(), offsets),
        this.transforms
      )
    );
  }

  rotate(angles: [number, number, number]): Group {
    const yaw = angles[2];
    const pitch = angles[1];
    const roll = angles[0];

    return new Group(
      this.children,
      mat4.multiply(
        mat4.create(),
        mat4.fromTaitBryanRotation(mat4.create(), yaw, pitch, roll),
        this.transforms
      )
    );
  }

  scale(factors: [number, number, number]): Group {
    return new Group(
      this.children,
      mat4.multiply(
        mat4.create(),
        mat4.fromScaling(mat4.create(), factors),
        this.transforms
      )
    );
  }

  toReplString(): string {
    return '<Group>';
  }

  ungroup(): Operable[] {
    // Return all children, but we need to account for this Group's unresolved
    // transforms by applying them to each child
    return this.children.map((child: Operable) => child.applyTransforms(this.transforms));
  }
}

export class Shape implements Operable, ReplResult {
  constructor(public solid: Solid = geom3.create()) {}

  applyTransforms(newTransforms: Mat4): Operable {
    // Return a new object for statelessness
    return new Shape(_transform(newTransforms, this.solid));
  }

  store(newTransforms: Mat4 = mat4.create()): void {
    Core.getRenderGroupManager()
      .storeShape(
        this.applyTransforms(newTransforms) as Shape
      );
  }

  translate(offsets: [number, number, number]): Shape {
    return new Shape(_translate(offsets, this.solid));
  }

  rotate(angles: [number, number, number]): Shape {
    return new Shape(_rotate(angles, this.solid));
  }

  scale(factors: [number, number, number]): Shape {
    return new Shape(_scale(factors, this.solid));
  }

  toReplString(): string {
    return '<Shape>';
  }
}

export class RenderGroup implements ReplResult {
  render: boolean = false;
  hasGrid: boolean = true;
  hasAxis: boolean = true;

  shapes: Shape[] = [];

  constructor(public canvasNumber: number) {}

  toReplString(): string {
    return `<Render #${this.canvasNumber}>`;
  }
}

export class RenderGroupManager {
  private canvasTracker: number = 1;
  private renderGroups: RenderGroup[] = [];

  constructor() {
    this.addRenderGroup();
  }

  private addRenderGroup() {
    // Passes in canvasTracker as is, then increments it
    this.renderGroups.push(new RenderGroup(this.canvasTracker++));
  }

  private getCurrentRenderGroup(): RenderGroup {
    return this.renderGroups.at(-1) as RenderGroup;
  }

  // Returns the old render group
  nextRenderGroup(
    oldHasGrid: boolean = false,
    oldHasAxis: boolean = false
  ): RenderGroup {
    const oldRenderGroup: RenderGroup = this.getCurrentRenderGroup();
    oldRenderGroup.render = true;
    oldRenderGroup.hasGrid = oldHasGrid;
    oldRenderGroup.hasAxis = oldHasAxis;

    this.addRenderGroup();

    return oldRenderGroup;
  }

  storeShape(shape: Shape) {
    this.getCurrentRenderGroup().shapes.push(shape);
  }

  shouldRender(): boolean {
    return this.getGroupsToRender().length > 0;
  }

  getGroupsToRender(): RenderGroup[] {
    return this.renderGroups.filter(
      (renderGroup: RenderGroup) => renderGroup.render
    );
  }
}

export class CsgModuleState {
  private componentCounter: number = 0;

  public readonly renderGroupManager: RenderGroupManager;

  public constructor() {
    this.renderGroupManager = new RenderGroupManager();
  }

  // Returns the new component number
  public nextComponent(): number {
    return ++this.componentCounter;
  }
}

export function centerPrimitive(shape: Shape) {
  // Move centre of Shape to 0.5, 0.5, 0.5
  const solid: Solid = _center(
    {
      relativeTo: [0.5, 0.5, 0.5]
    },
    shape.solid
  );
  return new Shape(solid);
}

export function hexToColor(hex: string): Color {
  const regex: RegExp
    = /^#?(?<red>[\da-f]{2})(?<green>[\da-f]{2})(?<blue>[\da-f]{2})$/iu;
  const potentialGroups: { [key: string]: string } | undefined
    = hex.match(regex)?.groups;
  if (potentialGroups === undefined) return [0, 0, 0];
  const groups: { [key: string]: string } = potentialGroups;

  return [
    parseInt(groups.red, 16) / 0xff,
    parseInt(groups.green, 16) / 0xff,
    parseInt(groups.blue, 16) / 0xff
  ];
}

export function colorToAlphaColor(
  color: Color,
  opacity: number = 1
): AlphaColor {
  return [...color, opacity];
}

export function hexToAlphaColor(hex: string): AlphaColor {
  return colorToAlphaColor(hexToColor(hex));
}
