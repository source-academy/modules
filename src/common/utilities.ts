/* [Exports] */
export function degreesToRadians(degrees: number): number {
  return (degrees / 360) * (2 * Math.PI);
}
