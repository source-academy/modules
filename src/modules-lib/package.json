{
  "name": "@sourceacademy/modules-lib",
  "description": "Common library used by Source Academy bundles and tabs",
  "private": true,
  "type": "module",
  "devDependencies": {
    "@types/react": "^18.3.1",
    "@types/react-dom": "^18.3.1",
    "jest": "^29.7.0",
    "ts-jest": "^29.1.2",
    "typescript": "^5.8.2"
  },
  "exports": {
    "./*": "./dist/*.js",
    "./types": "./dist/types/index.js"
  },
  "dependencies": {
    "@blueprintjs/core": "^5.10.2",
    "@blueprintjs/icons": "^5.9.0",
    "js-slang": "^1.0.81",
    "react": "^18.3.1",
    "react-dom": "^18.3.1"
  },
  "scripts": {
    "build": "tsc --project ./tsconfig.prod.json",
    "tsc": "tsc --project ./tsconfig.json",
    "test": "jest"
  },
  "jest": {
    "displayName": "Modules Library",
    "extensionsToTreatAsEsm": [
      ".ts"
    ],
    "preset": "ts-jest/presets/default-esm",
    "testMatch": [
      "<rootDir>/src/**/__tests__/**/*.ts"
    ]
  }
}
