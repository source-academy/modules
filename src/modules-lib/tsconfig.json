// modules-lib with tests
{
  "compilerOptions": {
    "noEmit": true
  },
  "extends": ["../tsconfig.json", "../tabs/tsconfig.json"],
  "include": ["./src"]
}