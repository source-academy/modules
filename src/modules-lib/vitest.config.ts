// Modules-lib test config
import { defineProject, mergeConfig } from 'vitest/config';
import rootConfig from '../../vitest.config';

export default mergeConfig(
  rootConfig,
  defineProject({
    test: {
      name: 'Modules Library',
      globals: true
    }
  })
);
