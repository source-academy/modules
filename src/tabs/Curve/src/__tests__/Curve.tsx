/* eslint-disable @typescript-eslint/no-unused-vars */
import { animate_3D_curve, animate_curve, draw_3D_connected, draw_connected } from '@sourceacademy/bundle-curve';
import type { CurveModuleState } from '@sourceacademy/bundle-curve/types';
import { mockDebuggerContext } from '@sourceacademy/modules-lib/utilities';
import { expect, test } from 'vitest';
import { CurveTab } from '..';

test('Curve animations error gracefully', () => {
  const badAnimation = animate_curve(1, 60, draw_connected(200), t => 1 as any);
  const mockContext = mockDebuggerContext<CurveModuleState>({ drawnCurves: [badAnimation] }, 'curve');
  expect(<CurveTab context={mockContext} />)
    .toMatchSnapshot();
});

test('Curve 3D animations error gracefully', () => {
  const badAnimation = animate_3D_curve(1, 60, draw_3D_connected(200), t => 1 as any);
  const mockContext = mockDebuggerContext<CurveModuleState>({ drawnCurves: [badAnimation] }, 'curve');
  expect(<CurveTab context={mockContext} />)
    .toMatchSnapshot();
});
