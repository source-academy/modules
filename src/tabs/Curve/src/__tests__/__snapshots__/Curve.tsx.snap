// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Curve 3D animations error gracefully 1`] = `
<CurveTab
  context={
    {
      "context": {
        "curve": {
          "state": {
            "drawnCurves": [
              AnimatedCurve {
                "angle": 0,
                "drawer": [Function],
                "duration": 1,
                "fps": 60,
                "func": [Function],
                "is3D": true,
                "toReplString": [Function],
              },
            ],
          },
        },
      },
    }
  }
/>
`;

exports[`Curve animations error gracefully 1`] = `
<CurveTab
  context={
    {
      "context": {
        "curve": {
          "state": {
            "drawnCurves": [
              AnimatedCurve {
                "angle": 0,
                "drawer": [Function],
                "duration": 1,
                "fps": 60,
                "func": [Function],
                "is3D": false,
                "toReplString": [Function],
              },
            ],
          },
        },
      },
    }
  }
/>
`;
